### PR TITLE
refactor regex

### LINKED ISSUE #10969: Corrupt source map paths for context module that uses negative lookahead
# Bug report

**What is the current behavior?**

Given a context module that uses a negative lookahead, e.g.

```js
// registers all files other than `./index.js`
const context = require.context( ".", true, /^(?!.*[\\/]index\.js$).*$/i, "lazy" );
```

Webpack generates a corrupted source map filenames for this context module.

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Create a context module as per the above example, on some deeply rooted path within your project structure, e.g. `/foo/bar/index.js`
2. Compile with source maps enabled.
3. Look at output in browser dev tools or open up the generated sourcemap.
3. Note the sourcemap has the context module at `webpack:///.*[\\/]index\.js$).*$`, instead of where it should go: `webpack:///foo/bar/index.js lazy ^(?!.*[\\/]index\.js$).*$`

**What is the expected behavior?**

For Webpack to generate the correct source map path:
`webpack:///foo/bar/index.js lazy ^(?!.*[\\/]index\.js$).*$`

**Other relevant information:**
webpack version:  4.43
Node.js version: 12.16.3
Operating System: Win 10 Pro 1909
Additional tools: -

### PR DESCRIPTION
**Pull request by Adityaperiwal**
resolves #10969

**What kind of change does this PR introduce?**
Bugfix

**Did you add tests for your changes?**
Yes

**Does this PR introduce a breaking change?**

No

**What needs to be documented once your changes are merged?**

Nothing

### CODE REVIEW

**Review by vankop — CHANGES_REQUESTED** [review]
Please add a test case that covers this case.

Take a look on `require.context` test cases https://github.com/webpack/webpack/tree/master/test/cases/context

File: lib/ContextModule.js [hunk 1]
```diff
@@ -144,7 +144,7 @@ class ContextModule extends Module {
 	prettyRegExp(regexString) {
 		// remove the "/" at the front and the beginning
 		// "/foo/" -> "foo"
-		return regexString.substring(1, regexString.length - 1);
+		return regexString.substring(1, regexString.length - 1).replace("!", "%21");
```
Comment (sokra) [inline lib/ContextModule.js:147]:
```suggestion
		return regexString.substring(1, regexString.length - 1).replace(/!/g, "%21");
```

**Review by sokra — CHANGES_REQUESTED** [review]
Please add a test case

File: lib/ContextModule.js (hunk 1 above)
Comment (Adityaperiwal) [reply lib/ContextModule.js:147]: > Please add a test case that covers this case.
> 
> Take a look on `require.context` test cases https://github.com/webpack/webpack/tree/master/test/cases/context

@sokra / @vankop `require.context()` just export three members `keys()`, `id` and `resolve()`, how can I get the `sourcemap` in tests?.

**Review by sokra — APPROVED** [review]

**Review by snitin315 — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Pull request merged**